Checking for presence of O_DIRECTORY and O_CLOEXEC.

Trying to build GAP on [ARCHER](https://www.archer.ac.uk/) with @raemarina @IrynaRaievska and @marioa, and observing this:

```
src/iostream.c: In function 'StartChildProcess':
src/iostream.c:396: error: 'O_DIRECTORY' undeclared (first use in this function)
src/iostream.c:396: error: (Each undeclared identifier is reported only once
src/iostream.c:396: error: for each function it appears in.)
src/iostream.c:396: error: 'O_CLOEXEC' undeclared (first use in this function)
make: *** [obj/src/iostream.lo] Error 1
```

As @fingolfin said, that means that default glibc is more than 10 years old and 
>  that's a regression we should fix in GAP: we detect the presence of `posix_spawn` and want to use it, but don't check for presence of `O_DIRECTORY` and `O_CLOEXEC`. .. In the meantime, a workaround might be to remove the test for `posix_spawn` from `configure.ac`.

module load autotools

autoconf --version
autoconf (GNU Autoconf) 2.69

autoconf and then make

We get:

obj/src/.libs/profile.o: In function `InformProfilingThatThisIsAForkedGAP':
/home/d155/d155/maryna/gap-4.11.0/src/profile.c:390: undefined reference to `endsWithgz'
collect2: ld returned 1 exit status
make: *** [gap] Error 1

--- a/src/iostream.c
+++ b/src/iostream.c
@@ -316,8 +316,22 @@ static Obj FuncDEFAULT_SIGCHLD_HANDLER(Obj self)
 #undef HAVE_POSIX_SPAWN
 #endif
 
+// if neither posix_spawn_file_actions_addchdir nor O_CLOEXEC are available,
+// our posix_spawn code should not be used. This affects e.g. Linux systems
+// with an old glibc, see https://github.com/gap-system/gap/issues/3918.
+#if !defined(HAVE_POSIX_SPAWN_FILE_ACTIONS_ADDCHDIR) && !defined(O_CLOEXEC)
+#undef HAVE_POSIX_SPAWN
+#endif
+
 
 #ifdef HAVE_POSIX_SPAWN
+
+// O_DIRECTORY is not available in old Linux / glibc versions, but the way we
+// use it below, it is optional anyway
+#ifndef O_DIRECTORY
+#define O_DIRECTORY 0
+#endif
+
 
 // The following function behaves like posix_spawn, but also
 // changes the working directory temporarily to 'dir'.

--- a/src/profile.c
+++ b/src/profile.c
@@ -334,13 +334,11 @@ static void leaveFunction(Obj func)
 ** of GAP's execution, before anything else is done.
 */
 
-#ifdef HAVE_POPEN
 static BOOL endsWithgz(const char * s)
 {
   s = strrchr(s, '.');
   return s && streq(s, ".gz");
 }
-#endif
 
 static void fopenMaybeCompressed(const char* name, struct ProfileState* ps)
 {

